nonthermal.cc

```diff
@@ -2692,15 +2750,46 @@ void nt_MPI_Bcast(const int modelgridindex, const int root) {
     }
 
     const auto frac_excitations_list_size = nt_solution[modelgridindex].frac_excitations_list.size();
+    int position = 0;
+
+    MPI_Barrier(MPI_COMM_WORLD);
+    if (root == my_rank) {
+      for (size_t excitationindex = 0; excitationindex < frac_excitations_list_size; excitationindex++) {
+        MPI_Pack(&nt_solution[modelgridindex].frac_excitations_list[excitationindex].frac_deposition, 1, MPI_DOUBLE,
+                 nt_maxstored_buffer, buffer_size, &position, MPI_COMM_WORLD);
+        MPI_Pack(&nt_solution[modelgridindex].frac_excitations_list[excitationindex].ratecoeffperdeposition, 1,
+                 MPI_DOUBLE, nt_maxstored_buffer, buffer_size, &position, MPI_COMM_WORLD);
+        MPI_Pack(&nt_solution[modelgridindex].frac_excitations_list[excitationindex].lineindex, 1, MPI_INT,
+                 nt_maxstored_buffer, buffer_size, &position, MPI_COMM_WORLD);
+      }
+    }
+
+    MPI_Barrier(MPI_COMM_WORLD);
+
+    if (modelgridindex % 1000 == 0) {
+      printout("Broadcasting for MGI = %d\n", modelgridindex);
+    }
+
+    MPI_Bcast(nt_maxstored_buffer, buffer_size, MPI_PACKED, root, MPI_COMM_WORLD);
+
+    MPI_Barrier(MPI_COMM_WORLD);
+
+    position = 0;
+
     for (size_t excitationindex = 0; excitationindex < frac_excitations_list_size; excitationindex++) {
```

Comment:
It would be nice if we could eliminate the complexity of the MPI_Packing and extra buffer after the nt_excitations list has been de-duplicated on the nodes. Check whether this is really necessary.

Reply:
Theoretically if the above is implemented, we could just reduce the quantity across each rank to 0 instead of casting.